### PR TITLE
Dev timeseries

### PR DESCRIPTION
**Pull request by cyildirim**
|      Q       |   A
|------------- | -----------
| Type         |feature
| BC Break     | yes/no
| Fixed issues |

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php [hunk 1]
```diff
@@ -420,6 +420,34 @@
...
+     * @var int|null
+     */
+    public $collectionSize;
+
+    /**
+     * READ-ONLY: If the collection is fixed size, the maximum number of elements to store in the collection.
+     *
+     * @var int|null
+     */
+    public $collectionMax;
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php:421]: There already is a `$collectionMax` property a few lines above.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 2]
```diff
@@ -65,7 +65,7 @@ public function __construct(protected DocumentManager $dm, protected ClassMetada
     public function ensureIndexes(?int $maxTimeMs = null, ?WriteConcern $writeConcern = null, bool $background = false): void
     {
         foreach ($this->metadataFactory->getAllMetadata() as $class) {
-            if ($class->isMappedSuperclass || $class->isEmbeddedDocument || $class->isQueryResultDocument || $class->isView()) {
+            if ($class->isMappedSuperclass || $class->isEmbeddedDocument || $class->isQueryResultDocument || $class->isView() || $class->isTimeSeriesDocument) {
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: MongoDB 6.0 adds support for secondary indexes on time series collections: https://www.mongodb.com/docs/manual/core/timeseries/timeseries-limitations/#time-series-secondary-indexes

With that in mind, I don't think this should be added.

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php [hunk 3]
```diff
@@ -179,6 +179,9 @@ public function loadMetadataForClass($className, \Doctrine\Persistence\Mapping\C
 
         if (isset($documentAnnot->collection)) {
             $metadata->setCollection($documentAnnot->collection);
+            if (is_array($documentAnnot->collection) && array_key_exists('timeseries', $documentAnnot->collection)) {
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php:182]: There should be a corresponding change in the XML metadata driver.

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php [hunk 4]
```diff
@@ -420,6 +420,34 @@
      */
     public $collectionMax;
 
+        /**
+     * READ-ONLY: If the collection is fixed size, its size in bytes.
+     *
+     * @var int|null
+     */
+    public $collectionSize;
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php]: Public properties are a leftover from old PHP versions where method calls were associated with a significant performance penalty. Nowadays there is no problem to hide these values in private properties and expose them via getters if necessary.

Comment (cyildirim) [reply lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php]: I've removed `$collectionSize` here as there was already one. Should I convert to current ones to private ?

**Comment by bradroc** [comment]
@cyildirim this is looking great. Thanks for your contribution 🙏

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php (hunk 4 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php]: Sorry for the delay. Yes, any new properties that are introduced should be made private so we have fewer breaks when we eventually make everything private.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 5]
```diff
@@ -339,7 +339,7 @@ public function updateValidators(?int $maxTimeMs = null, ?WriteConcern $writeCon
     public function updateDocumentValidator(string $documentName, ?int $maxTimeMs = null, ?WriteConcern $writeConcern = null): void
     {
         $class = $this->dm->getClassMetadata($documentName);
-        if ($class->isMappedSuperclass || $class->isEmbeddedDocument || $class->isQueryResultDocument || $class->isView() || $class->isFile) {
+        if ($class->isMappedSuperclass || $class->isEmbeddedDocument || $class->isQueryResultDocument || $class->isView() || $class->isFile || $class->isTimeSeriesDocument) {
             throw new InvalidArgumentException('Cannot update validators for files, views, mapped super classes, embedded documents or aggregation result documents.');
```
Comment (lstrojny) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php:343]: Message needs to be changed as well

**Review by lstrojny — COMMENTED** [review]
Tests should be added for ClassMetadata and XmlDriver.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php (hunk 5 above)
Comment (cyildirim) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php:343]: Thanks for reviewing. I've fixed the tests for now and started to write some test for this bit.

**Comment by alcaeus** [comment]
@cyildirim Your work was picked up by @PReimers recently and continued in #2597. He's also started writing tests, so maybe you two can get together and see what else is missing?

**Comment by alcaeus** [comment]
Closing as #2687 has been merged, containing the feature.

**Pull request closed without merging**